File: backend/src/main/java/com/repopilot/knowledge/KnowledgeUtils.java
package com.repopilot.knowledge;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.*;
import java.util.regex.Pattern;

public final class KnowledgeUtils {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Pattern TOKEN_PATTERN = Pattern.compile("[a-zA-Z_][a-zA-Z0-9_./-]*|[\\u4e00-\\u9fff]{2,}");

    private KnowledgeUtils() {}

    public static boolean shouldSkipPath(String path) {
        String[] parts = path.split("/");
        for (int i = 0; i < parts.length - 1; i++) {
            if (KnowledgePolicy.SKIP_DIRS.contains(parts[i]) || parts[i].startsWith(".")) {
                return true;
            }
        }
        return false;
    }

    public static boolean isTextFile(String path) {
        String lower = path.toLowerCase();
        if (lower.endsWith("dockerfile")) {
            return true;
        }
        for (String ext : KnowledgePolicy.TEXT_EXTENSIONS) {
            if (lower.endsWith(ext)) {
                return true;
            }
        }
        return false;
    }

    public static int filePriority(String path, int size) {
        String lower = path.toLowerCase();
        String name = lower.contains("/") ? lower.substring(lower.lastIndexOf('/') + 1) : lower;
        int rank;
        if (name.startsWith("readme")) {
            rank = 0;
        } else if (Set.of("package.json", "requirements.txt", "main.py", "main.tsx").contains(name)) {
            rank = 1;
        } else if (lower.startsWith("src/") || lower.startsWith("backend/") || lower.startsWith("app/")) {
            rank = 2;
        } else if (lower.startsWith("docs/")) {
            rank = 3;
        } else {
            rank = 4;
        }
        return rank * 1_000_000 + size;
    }

    public static String detectLanguage(String path) {
        Map<String, String> mapping = Map.of(
                ".py", "Python", ".ts", "TypeScript", ".tsx", "TypeScript",
                ".js", "JavaScript", ".jsx", "JavaScript", ".md", "Markdown",
                ".json", "JSON", ".java", "Java", ".go", "Go", ".rs", "Rust"
        );
        String lower = path.toLowerCase();
        for (var entry : mapping.entrySet()) {
            if (lower.endsWith(entry.getKey())) {
                return entry.getValue();
            }
        }
        return null;
    }

    public static List<Map<String, Object>> chunkText(String content, String filePath) {
        List<Map<String, Object>> chunks = new ArrayList<>();
        if (content == null || content.isBlank()) {
            return chunks;
        }
        int start = 0;
        int chunkIndex = 0;
        while (start < content.length()) {
            int end = Math.min(content.length(), start + KnowledgePolicy.CHUNK_SIZE);
            String piece = content.substring(start, end);
            int lineNo = content.substring(0, start).split("\n", -1).length;
            chunks.add(Map.of(
                    "file_path", filePath,
                    "chunk_index", chunkIndex,
                    "content", piece,
                    "start_line", lineNo
            ));
            chunkIndex++;
            if (end >= content.length()) {
                break;
            }
            start = Math.max(end - KnowledgePolicy.CHUNK_OVERLAP, start + 1);
        }
        return chunks;
    }

    public static List<Map<String, Object>> buildTree(List<String> paths) {
        Map<String, Object> root = new TreeMap<>();
        for (String path : paths.stream().sorted().toList()) {
            String[] parts = path.split("/");
            Map<String, Object> cursor = root;
            for (int i = 0; i < parts.length; i++) {
                String part = parts[i];
                @SuppressWarnings("unchecked")
                Map<String, Object> node = (Map<String, Object>) cursor.computeIfAbsent(part, k -> new LinkedHashMap<>());
                node.putIfAbsent("__path__", String.join("/", Arrays.copyOfRange(parts, 0, i + 1)));
                if (i == parts.length - 1) {
                    node.put("__is_file__", true);
                }
                @SuppressWarnings("unchecked")
                Map<String, Object> children = (Map<String, Object>) node.computeIfAbsent("__children__", k -> new LinkedHashMap<>());
                cursor = children;
            }
        }

        List<Map<String, Object>> nodes = new ArrayList<>();
        for (var entry : root.entrySet()) {
            nodes.add(toNode((Map<String, Object>) entry.getValue(), entry.getKey()));
        }
        return nodes;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> toNode(Map<String, Object> node, String name) {
        boolean isFile = Boolean.TRUE.equals(node.get("__is_file__"));
        String fullPath = (String) node.getOrDefault("__path__", name);
        Map<String, Object> childrenDict = (Map<String, Object>) node.getOrDefault("__children__", Map.of());
        List<Map<String, Object>> children = new ArrayList<>();
        for (var entry : childrenDict.entrySet()) {
            children.add(toNode((Map<String, Object>) entry.getValue(), entry.getKey()));
        }
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("key", fullPath);
        result.put("title", name);
        result.put("type", isFile ? "file" : (fullPath.split("/").length <= 1 ? "folder" : "module"));
        result.put("children", children.isEmpty() ? null : children);
        return result;
    }

    public static List<Map<String, Object>> extractModules(List<FileRow> files) {
        Map<String, List<FileRow>> buckets = new TreeMap<>();
        for (FileRow file : files) {
            if (!"file".equals(file.fileType())) {
                continue;
            }
            String top = file.path().contains("/") ? file.path().split("/")[0] : file.path();
            buckets.computeIfAbsent(top, k -> new ArrayList<>()).add(file);
        }
        List<Map<String, Object>> modules = new ArrayList<>();
        for (var entry : buckets.entrySet()) {
            Set<String> langs = new TreeSet<>();
            for (FileRow row : entry.getValue()) {
                if (row.language() != null) {
                    langs.add(row.language());
                }
            }
            modules.add(Map.of(
                    "name", entry.getKey(),
                    "desc", "包含 " + entry.getValue().size() + " 个已索引文件",
                    "files", entry.getValue().size(),
                    "deps", langs.stream().limit(5).toList()
            ));
            if (modules.size() >= 12) {
                break;
            }
        }
        return modules;
    }

    public static Map<String, Integer> extractLanguageStats(List<FileRow> files) {
        Map<String, Integer> stats = new HashMap<>();
        for (FileRow file : files) {
            if (!"file".equals(file.fileType())) {
                continue;
            }
            String lang = file.language() == null ? "Other" : file.language();
            stats.merge(lang, 1, Integer::sum);
        }
        return stats.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .collect(LinkedHashMap::new, (m, e) -> m.put(e.getKey(), e.getValue()), LinkedHashMap::putAll);
    }

    public static String extractRepoSummary(String fullName, Map<String, String> fileMap, List<Map<String, Object>> modules) {
        Optional<String> readme = fileMap.entrySet().stream()
                .filter(e -> e.getKey().toLowerCase().endsWith("readme.md") || e.getKey().toLowerCase().contains("readme"))
                .map(Map.Entry::getValue)
                .findFirst();
        if (readme.isPresent()) {
            String preview = readme.get().replaceAll("\\s+", " ").trim();
            if (preview.length() > 360) {
                preview = preview.substring(0, 360);
            }
            return fullName + "：" + preview;
        }
        if (!modules.isEmpty()) {
            String names = modules.stream().limit(6).map(m -> (String) m.get("name")).reduce((a, b) -> a + "、" + b).orElse("");
            return fullName + " 已索引主要模块：" + names + "。";
        }
        return fullName + " 已完成基础目录与文件索引。";
    }

    public static List<String> extractDependencies(Map<String, String> fileMap) {
        Set<String> deps = new TreeSet<>();
        String packageJson = fileMap.get("package.json");
        if (packageJson != null) {
            try {
                JsonNode data = MAPPER.readTree(packageJson);
                for (String key : List.of("dependencies", "devDependencies")) {
                    JsonNode node = data.path(key);
                    if (node.isObject()) {
                        node.fieldNames().forEachRemaining(deps::add);
                    }
                }
            } catch (Exception ignored) {
            }
        }
        String requirements = fileMap.getOrDefault("requirements.txt", fileMap.get("backend/requirements.txt"));
        if (requirements != null) {
            for (String line : requirements.split("\n")) {
                line = line.trim();
                if (!line.isEmpty() && !line.startsWith("#")) {
                    deps.add(line.split("[<>=!]")[0].trim());
                }
            }
        }
        return deps.stream().limit(20).toList();
    }

    public static String classifyQuestion(String question) {
        String q = question.toLowerCase();
        if (q.matches(".*(什么|是什么|what).*")) {
            return "what";
        }
        if (q.matches(".*(哪里|在哪|where|路径|文件).*")) {
            return "where";
        }
        if (q.matches(".*(如何|怎么|怎样|how).*")) {
            return "how";
        }
        return "what";
    }

    public static List<String> tokenize(String question) {
        Set<String> stopWords = Set.of("什么", "如何", "怎么", "哪里", "项目", "代码", "the", "and", "for");
        var matcher = TOKEN_PATTERN.matcher(question.toLowerCase());
        List<String> tokens = new ArrayList<>();
        while (matcher.find()) {
            String token = matcher.group();
            if (!stopWords.contains(token) && token.length() > 1) {
                tokens.add(token);
            }
        }
        return tokens.stream().distinct().toList();
    }

    public record FileRow(String path, String fileType, int size, String language, String content) {}
}

File: prototype/backend/knowledge_utils.py
import json
import re
from dataclasses import dataclass

SKIP_DIRS = {
    "node_modules",
    ".git",
    "dist",
    "build",
    ".next",
    "__pycache__",
    ".venv",
    "venv",
    "coverage",
    ".idea",
    ".vscode",
}

PRIORITY_FILENAMES = {
    "readme.md",
    "package.json",
    "requirements.txt",
    "pyproject.toml",
    "vite.config.ts",
    "vite.config.js",
    "tsconfig.json",
    "src/main.tsx",
    "src/main.ts",
    "src/app.tsx",
    "src/app.ts",
    "main.py",
    "app.py",
}

TEXT_EXTENSIONS = {
    ".md",
    ".txt",
    ".py",
    ".ts",
    ".tsx",
    ".js",
    ".jsx",
    ".json",
    ".yml",
    ".yaml",
    ".toml",
    ".html",
    ".css",
    ".scss",
    ".vue",
    ".go",
    ".rs",
    ".java",
    ".kt",
    ".sql",
    ".sh",
    ".env.example",
}

MAX_FILES = 180
MAX_FILE_BYTES = 80_000
CHUNK_SIZE = 900
CHUNK_OVERLAP = 120


@dataclass
class FileRecord:
    path: str
    file_type: str
    size: int
    language: str | None
    content: str | None = None


def should_skip_path(path: str) -> bool:
    parts = path.split("/")
    return any(part in SKIP_DIRS or part.startswith(".") for part in parts[:-1])


def is_text_file(path: str) -> bool:
    lower = path.lower()
    return any(lower.endswith(ext) for ext in TEXT_EXTENSIONS) or lower.endswith("dockerfile")


def file_priority(path: str, size: int) -> tuple[int, int, str]:
    lower = path.lower()
    name = lower.rsplit("/", 1)[-1]
    if name.startswith("readme"):
        rank = 0
    elif lower in PRIORITY_FILENAMES or name in PRIORITY_FILENAMES:
        rank = 1
    elif lower.startswith(("src/", "app/", "pages/", "components/", "backend/")):
        rank = 2
    elif lower.startswith(("docs/", "documentation/")):
        rank = 3
    else:
        rank = 4
    return (rank, size, lower)


def detect_language(path: str) -> str | None:
    mapping = {
        ".py": "Python",
        ".ts": "TypeScript",
        ".tsx": "TypeScript",
        ".js": "JavaScript",
        ".jsx": "JavaScript",
        ".md": "Markdown",
        ".json": "JSON",
        ".go": "Go",
        ".rs": "Rust",
        ".java": "Java",
        ".vue": "Vue",
        ".css": "CSS",
        ".html": "HTML",
    }
    for ext, lang in mapping.items():
        if path.lower().endswith(ext):
            return lang
    return None


def short_preview(content: str, max_len: int = 260) -> str:
    text = re.sub(r"\s+", " ", content).strip()
    return text[:max_len]


def chunk_text(content: str, file_path: str) -> list[dict]:
    lines = content.splitlines()
    chunks: list[dict] = []
    if not content.strip():
        return chunks

    start = 0
    chunk_index = 0
    while start < len(content):
        end = min(len(content), start + CHUNK_SIZE)
        piece = content[start:end]
        line_no = content[:start].count("\n") + 1
        chunks.append(
            {
                "file_path": file_path,
                "chunk_index": chunk_index,
                "content": piece,
                "start_line": line_no,
            }
        )
        chunk_index += 1
        if end >= len(content):
            break
        start = max(end - CHUNK_OVERLAP, start + 1)
    return chunks


def build_tree(paths: list[str]) -> list[dict]:
    root: dict = {}

    for path in sorted(paths):
        parts = path.split("/")
        cursor = root
        for idx, part in enumerate(parts):
            cursor.setdefault(part, {"__children__": {}, "__path__": "/".join(parts[: idx + 1])})
            if idx == len(parts) - 1:
                cursor[part]["__is_file__"] = True
            cursor = cursor[part]["__children__"]

    def to_nodes(node: dict, name: str, full_path: str) -> dict:
        is_file = node.get("__is_file__", False)
        children_dict = node.get("__children__", {})
        children = [
            to_nodes(child, child_name, child.get("__path__", f"{full_path}/{child_name}"))
            for child_name, child in children_dict.items()
        ]
        return {
            "key": full_path,
            "title": name,
            "type": "file" if is_file else ("folder" if full_path.count("/") <= 1 else "module"),
            "children": children or None,
        }

    return [to_nodes(root[name], name, name) for name in root]


def extract_modules(files: list[FileRecord]) -> list[dict]:
    buckets: dict[str, list[FileRecord]] = {}
    for file in files:
        if file.file_type != "file":
            continue
        top = file.path.split("/")[0]
        buckets.setdefault(top, []).append(file)

    modules = []
    for name, items in sorted(buckets.items()):
        langs = sorted({f.language for f in items if f.language})
        modules.append(
            {
                "name": name,
                "desc": f"包含 {len(items)} 个已索引文件",
                "files": len(items),
                "deps": langs[:5],
            }
        )
    return modules[:12]


def extract_language_stats(files: list[FileRecord]) -> dict[str, int]:
    stats: dict[str, int] = {}
    for file in files:
        if file.file_type != "file":
            continue
        lang = file.language or "Other"
        stats[lang] = stats.get(lang, 0) + 1
    return dict(sorted(stats.items(), key=lambda item: item[1], reverse=True))


def extract_repo_summary(full_name: str, file_map: dict[str, str], modules: list[dict]) -> str:
    readme = next((content for path, content in file_map.items() if path.lower().rsplit("/", 1)[-1].startswith("readme")), "")
    if readme:
        first_heading = next((line.strip("# ").strip() for line in readme.splitlines() if line.startswith("#")), "")
        preview = short_preview(readme, 360)
        return f"{full_name}：{first_heading or 'README 摘要'}。{preview}"
    if modules:
        names = "、".join(module["name"] for module in modules[:6])
        return f"{full_name} 已索引主要模块：{names}。"
    return f"{full_name} 已完成基础目录与文件索引。"


def extract_dependencies(file_map: dict[str, str]) -> list[str]:
    deps: list[str] = []
    package_json = file_map.get("package.json")
    if package_json:
        try:
            data = json.loads(package_json)
            for key in ("dependencies", "devDependencies"):
                deps.extend(data.get(key, {}).keys())
        except json.JSONDecodeError:
            pass

    requirements = file_map.get("requirements.txt") or file_map.get("backend/requirements.txt")
    if requirements:
        for line in requirements.splitlines():
            line = line.strip()
            if line and not line.startswith("#"):
                deps.append(re.split(r"[<>=!]", line)[0].strip())

    return sorted(set(deps))[:20]
